File: tests/cpp/core/schema/test_pedals.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated Pedals FlatBuffer types.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/pedals_generated.h>
#include <schema/serialized.hpp>
#include <schema/timestamp_generated.h>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs for Generic3AxisPedalOutput table.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2

static_assert(core::Generic3AxisPedalOutput::VT_LEFT_PEDAL == VT(0));
static_assert(core::Generic3AxisPedalOutput::VT_RIGHT_PEDAL == VT(1));
static_assert(core::Generic3AxisPedalOutput::VT_RUDDER == VT(2));

static_assert(core::Generic3AxisPedalOutputRecord::VT_DATA == VT(0));
static_assert(core::Generic3AxisPedalOutputRecord::VT_TIMESTAMP == VT(1));

// =============================================================================
// Generic3AxisPedalOutputT Tests (table native type)
// =============================================================================
TEST_CASE("Generic3AxisPedalOutputT default construction", "[pedals][native]")
{
    core::Generic3AxisPedalOutputT output;

    // Float fields should be zero by default.
    CHECK(output.left_pedal == 0.0f);
    CHECK(output.right_pedal == 0.0f);
    CHECK(output.rudder == 0.0f);
}

TEST_CASE("Generic3AxisPedalOutputT can store pedal values", "[pedals][native]")
{
    core::Generic3AxisPedalOutputT output;

    output.left_pedal = 0.75f;
    output.right_pedal = 0.25f;
    output.rudder = 0.5f;

    CHECK(output.left_pedal == Catch::Approx(0.75f));
    CHECK(output.right_pedal == Catch::Approx(0.25f));
    CHECK(output.rudder == Catch::Approx(0.5f));
}

TEST_CASE("Generic3AxisPedalOutputT can store full output", "[pedals][native]")
{
    core::Generic3AxisPedalOutputT output;

    output.left_pedal = 1.0f;
    output.right_pedal = 0.0f;
    output.rudder = -0.5f;

    CHECK(output.left_pedal == Catch::Approx(1.0f));
    CHECK(output.right_pedal == Catch::Approx(0.0f));
    CHECK(output.rudder == Catch::Approx(-0.5f));
}

// =============================================================================
// Optionality Tests
//
// A tracker expresses "no pedal data" with an empty Serialized handle rather than a
// wrapper table holding a null payload, so these cover the handle's absent state.
// =============================================================================
TEST_CASE("Serialized pedal output is empty by default", "[pedals][tracked]")
{
    core::Serialized<core::Generic3AxisPedalOutput> data;

    CHECK_FALSE(static_cast<bool>(data));
    CHECK(data.get() == nullptr);
}

TEST_CASE("Serialized pedal output round-trips its fields", "[pedals][tracked]")
{
    core::Generic3AxisPedalOutputT native;
    native.left_pedal = 0.5f;
    native.right_pedal = 0.3f;
    native.rudder = -0.2f;

    const auto data = core::pack<core::Generic3AxisPedalOutput>(native);

    REQUIRE(static_cast<bool>(data));
    CHECK(data->left_pedal() == Catch::Approx(0.5f));
    CHECK(data->right_pedal() == Catch::Approx(0.3f));
    CHECK(data->rudder() == Catch::Approx(-0.2f));
}

TEST_CASE("Serialized pedal output can be returned to empty", "[pedals][tracked]")
{
    core::Generic3AxisPedalOutputT native;
    native.left_pedal = 0.8f;

    auto data = core::pack<core::Generic3AxisPedalOutput>(native);
    REQUIRE(static_cast<bool>(data));

    data = core::Serialized<core::Generic3AxisPedalOutput>();

    CHECK_FALSE(static_cast<bool>(data));
}

TEST_CASE("Generic3AxisPedalOutputRecord serialization with tracked data", "[pedals][tracked][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Create native object with data.
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.5f;
    output.right_pedal = 0.5f;
    output.rudder = 0.0f;

    // Pack data before starting the RecordBuilder to avoid nested table building.
    auto data_offset = core::Generic3AxisPedalOutput::Pack(builder, &output);
    core::Generic3AxisPedalOutputRecordBuilder record_builder(builder);
    record_builder.add_data(data_offset);
    builder.Finish(record_builder.Finish());

    // Deserialize.
    auto* record = flatbuffers::GetRoot<core::Generic3AxisPedalOutputRecord>(builder.GetBufferPointer());
    CHECK(record->data()->left_pedal() == Catch::Approx(0.5f));
    CHECK(record->data()->right_pedal() == Catch::Approx(0.5f));
    CHECK(record->data()->rudder() == Catch::Approx(0.0f));
}

TEST_CASE("Generic3AxisPedalOutputRecord serialization without data", "[pedals][tracked][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Serialize a record with no data (inactive pedal).
    core::Generic3AxisPedalOutputRecordBuilder record_builder(builder);
    builder.Finish(record_builder.Finish());

    // Deserialize.
    auto* record = flatbuffers::GetRoot<core::Generic3AxisPedalOutputRecord>(builder.GetBufferPointer());
    CHECK(record->data() == nullptr);
}

// =============================================================================
// Generic3AxisPedalOutput Serialization Tests
// =============================================================================
TEST_CASE("Generic3AxisPedalOutput serialization and deserialization", "[pedals][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Create native object.
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.8f;
    output.right_pedal = 0.2f;
    output.rudder = 0.33f;

    // Serialize.
    auto offset = core::Generic3AxisPedalOutput::Pack(builder, &output);
    builder.Finish(offset);

    // Deserialize.
    auto* deserialized = flatbuffers::GetRoot<core::Generic3AxisPedalOutput>(builder.GetBufferPointer());

    // Verify pedal values.
    CHECK(deserialized->left_pedal() == Catch::Approx(0.8f));
    CHECK(deserialized->right_pedal() == Catch::Approx(0.2f));
    CHECK(deserialized->rudder() == Catch::Approx(0.33f));
}

TEST_CASE("Generic3AxisPedalOutput can be unpacked from buffer", "[pedals][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Create native object with minimal data.
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.5f;

    // Serialize.
    auto offset = core::Generic3AxisPedalOutput::Pack(builder, &output);
    builder.Finish(offset);

    // Deserialize to table.
    auto* table = flatbuffers::GetRoot<core::Generic3AxisPedalOutput>(builder.GetBufferPointer());

    // Unpack to native.
    auto unpacked = std::make_unique<core::Generic3AxisPedalOutputT>();
    table->UnPackTo(unpacked.get());

    // Verify.
    CHECK(unpacked->left_pedal == Catch::Approx(0.5f));
}


// =============================================================================
// Realistic Foot Pedal Scenarios
// =============================================================================
TEST_CASE("Generic3AxisPedalOutput full forward press", "[pedals][scenario]")
{
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 1.0f;
    output.right_pedal = 1.0f;
    output.rudder = 0.0f;

    CHECK(output.left_pedal == Catch::Approx(1.0f));
    CHECK(output.right_pedal == Catch::Approx(1.0f));
    CHECK(output.rudder == Catch::Approx(0.0f));
}

TEST_CASE("Generic3AxisPedalOutput left turn with rudder", "[pedals][scenario]")
{
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.5f;
    output.right_pedal = 0.5f;
    output.rudder = -1.0f; // Full left rudder.

    CHECK(output.rudder == Catch::Approx(-1.0f));
}

TEST_CASE("Generic3AxisPedalOutput right turn with rudder", "[pedals][scenario]")
{
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.5f;
    output.right_pedal = 0.5f;
    output.rudder = 1.0f; // Full right rudder.

    CHECK(output.rudder == Catch::Approx(1.0f));
}

TEST_CASE("Generic3AxisPedalOutput differential braking", "[pedals][scenario]")
{
    // Simulating differential braking (left brake only).
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.0f; // Left brake applied.
    output.right_pedal = 0.8f; // Right pedal pressed.
    output.rudder = 0.0f;

    CHECK(output.left_pedal == Catch::Approx(0.0f));
    CHECK(output.right_pedal == Catch::Approx(0.8f));
}

TEST_CASE("Generic3AxisPedalOutput neutral position", "[pedals][scenario]")
{
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.0f;
    output.right_pedal = 0.0f;
    output.rudder = 0.0f;

    CHECK(output.left_pedal == Catch::Approx(0.0f));
    CHECK(output.right_pedal == Catch::Approx(0.0f));
    CHECK(output.rudder == Catch::Approx(0.0f));
}

// =============================================================================
// Edge Cases
// =============================================================================
TEST_CASE("Generic3AxisPedalOutput with negative values", "[pedals][edge]")
{
    // Although pedal values are typically 0-1, test negative values.
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = -0.5f;
    output.right_pedal = -0.25f;
    output.rudder = -1.0f;

    CHECK(output.left_pedal == Catch::Approx(-0.5f));
    CHECK(output.right_pedal == Catch::Approx(-0.25f));
    CHECK(output.rudder == Catch::Approx(-1.0f));
}

TEST_CASE("Generic3AxisPedalOutput with values greater than 1", "[pedals][edge]")
{
    // Test values exceeding typical range.
    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 1.5f;
    output.right_pedal = 2.0f;
    output.rudder = 1.5f;

    CHECK(output.left_pedal == Catch::Approx(1.5f));
    CHECK(output.right_pedal == Catch::Approx(2.0f));
    CHECK(output.rudder == Catch::Approx(1.5f));
}

TEST_CASE("Generic3AxisPedalOutput buffer size is reasonable", "[pedals][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    core::Generic3AxisPedalOutputT output;
    output.left_pedal = 0.0f;
    output.right_pedal = 0.0f;
    output.rudder = 0.0f;

    auto offset = core::Generic3AxisPedalOutput::Pack(builder, &output);
    builder.Finish(offset);

    // Buffer should be reasonably small (under 100 bytes for this simple message).
    CHECK(builder.GetSize() < 100);
}

// =============================================================================
// Generic3AxisPedalOutputRecord Tests (timestamp lives on the Record wrapper)
// =============================================================================
TEST_CASE("Generic3AxisPedalOutputRecord serialization with DeviceDataTimestamp", "[pedals][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    auto record = std::make_shared<core::Generic3AxisPedalOutputRecordT>();
    record->data = std::make_shared<core::Generic3AxisPedalOutputT>();
    record->data->left_pedal = 0.8f;
    record->data->right_pedal = 0.2f;
    record->data->rudder = 0.5f;
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1000000000LL, 2000000000LL, 3000000000LL);

    auto offset = core::Generic3AxisPedalOutputRecord::Pack(builder, record.get());
    builder.Finish(offset);

    auto deserialized = flatbuffers::GetRoot<core::Generic3AxisPedalOutputRecord>(builder.GetBufferPointer());

    CHECK(deserialized->timestamp()->available_time_local_common_clock() == 1000000000LL);
    CHECK(deserialized->timestamp()->sample_time_local_common_clock() == 2000000000LL);
    CHECK(deserialized->timestamp()->sample_time_raw_device_clock() == 3000000000LL);
    CHECK(deserialized->data()->left_pedal() == Catch::Approx(0.8f));
    CHECK(deserialized->data()->right_pedal() == Catch::Approx(0.2f));
    CHECK(deserialized->data()->rudder() == Catch::Approx(0.5f));
}

TEST_CASE("Generic3AxisPedalOutputRecord can be unpacked with DeviceDataTimestamp", "[pedals][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    auto original = std::make_shared<core::Generic3AxisPedalOutputRecordT>();
    original->data = std::make_shared<core::Generic3AxisPedalOutputT>();
    original->data->left_pedal = 0.5f;
    original->timestamp = std::make_shared<core::DeviceDataTimestamp>(111LL, 222LL, 333LL);

    auto offset = core::Generic3AxisPedalOutputRecord::Pack(builder, original.get());
    builder.Finish(offset);

    auto fb = flatbuffers::GetRoot<core::Generic3AxisPedalOutputRecord>(builder.GetBufferPointer());
    auto unpacked = std::make_shared<core::Generic3AxisPedalOutputRecordT>();
    fb->UnPackTo(unpacked.get());

    CHECK(unpacked->timestamp->available_time_local_common_clock() == 111LL);
    CHECK(unpacked->timestamp->sample_time_local_common_clock() == 222LL);
    CHECK(unpacked->timestamp->sample_time_raw_device_clock() == 333LL);
    CHECK(unpacked->data->left_pedal == Catch::Approx(0.5f));
}

File: tests/python/core/schema/test_pedals.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for Generic3AxisPedalOutput type in isaacteleop.schema.

Tests the following FlatBuffers types:
- Generic3AxisPedalOutput: Table with left_pedal, right_pedal, and rudder
- Generic3AxisPedalOutputRecord: Record wrapper carrying DeviceDataTimestamp

Timestamps are carried by Generic3AxisPedalOutputRecord, not Generic3AxisPedalOutput.
"""

import pytest

from isaacteleop.schema import (
    Generic3AxisPedalOutputRecord,
    Generic3AxisPedalOutput,
    DeviceDataTimestamp,
)


class TestGeneric3AxisPedalOutputConstruction:
    """Tests for Generic3AxisPedalOutput table construction."""

    def test_default_construction(self):
        """Test default construction creates Generic3AxisPedalOutput with default-initialized fields."""
        output = Generic3AxisPedalOutput()

        assert output.left_pedal == 0.0
        assert output.right_pedal == 0.0
        assert output.rudder == 0.0

    def test_repr(self):
        """Test __repr__ returns meaningful string."""
        output = Generic3AxisPedalOutput()
        repr_str = repr(output)

        assert "Generic3AxisPedalOutput" in repr_str


class TestGeneric3AxisPedalOutputPedals:
    """Tests that each pedal field round-trips through the encoding."""

    def test_left_pedal(self):
        """Test encoding left pedal value."""
        output = Generic3AxisPedalOutput(left_pedal=0.75)

        assert output.left_pedal == pytest.approx(0.75)

    def test_right_pedal(self):
        """Test encoding right pedal value."""
        output = Generic3AxisPedalOutput(right_pedal=0.5)

        assert output.right_pedal == pytest.approx(0.5)

    def test_rudder(self):
        """Test encoding rudder value."""
        output = Generic3AxisPedalOutput(rudder=-0.33)

        assert output.rudder == pytest.approx(-0.33)

    def test_all_pedal_values(self):
        """Test encoding all pedal values."""
        output = Generic3AxisPedalOutput(left_pedal=0.8, right_pedal=0.2, rudder=0.5)

        assert output.left_pedal == pytest.approx(0.8)
        assert output.right_pedal == pytest.approx(0.2)
        assert output.rudder == pytest.approx(0.5)


class TestGeneric3AxisPedalOutputCombined:
    """Tests for Generic3AxisPedalOutput with multiple fields set."""

    def test_full_output(self):
        """Test with all fields set."""
        output = Generic3AxisPedalOutput(left_pedal=1.0, right_pedal=0.0, rudder=-0.5)

        assert output.left_pedal == pytest.approx(1.0)
        assert output.right_pedal == pytest.approx(0.0)
        assert output.rudder == pytest.approx(-0.5)


class TestGeneric3AxisPedalOutputScenarios:
    """Tests for realistic foot pedal input scenarios."""

    def test_full_forward_press(self):
        """Test full forward press on both pedals."""
        output = Generic3AxisPedalOutput(left_pedal=1.0, right_pedal=1.0, rudder=0.0)

        assert output.left_pedal == pytest.approx(1.0)
        assert output.right_pedal == pytest.approx(1.0)
        assert output.rudder == pytest.approx(0.0)

    def test_left_turn_with_rudder(self):
        """Test left turn using rudder."""
        output = Generic3AxisPedalOutput(
            left_pedal=0.5, right_pedal=0.5, rudder=-1.0
        )  # Full left rudder.

        assert output.rudder == pytest.approx(-1.0)

    def test_right_turn_with_rudder(self):
        """Test right turn using rudder."""
        output = Generic3AxisPedalOutput(
            left_pedal=0.5, right_pedal=0.5, rudder=1.0
        )  # Full right rudder.

        assert output.rudder == pytest.approx(1.0)

    def test_differential_braking(self):
        """Test differential braking scenario."""
        # Left brake applied, right pedal pressed.
        output = Generic3AxisPedalOutput(left_pedal=0.0, right_pedal=0.8, rudder=0.0)

        assert output.left_pedal == pytest.approx(0.0)
        assert output.right_pedal == pytest.approx(0.8)

    def test_neutral_position(self):
        """Test neutral/idle position."""
        output = Generic3AxisPedalOutput()

        assert output.left_pedal == pytest.approx(0.0)
        assert output.right_pedal == pytest.approx(0.0)
        assert output.rudder == pytest.approx(0.0)


class TestGeneric3AxisPedalOutputEdgeCases:
    """Edge case tests for Generic3AxisPedalOutput table."""

    def test_negative_pedal_values(self):
        """Test with negative pedal values (edge case)."""
        output = Generic3AxisPedalOutput(
            left_pedal=-0.5, right_pedal=-0.25, rudder=-1.0
        )

        assert output.left_pedal == pytest.approx(-0.5)
        assert output.right_pedal == pytest.approx(-0.25)
        assert output.rudder == pytest.approx(-1.0)

    def test_values_greater_than_one(self):
        """Test with pedal values exceeding typical range."""
        output = Generic3AxisPedalOutput(left_pedal=1.5, right_pedal=2.0, rudder=1.5)

        assert output.left_pedal == pytest.approx(1.5)
        assert output.right_pedal == pytest.approx(2.0)
        assert output.rudder == pytest.approx(1.5)

    def test_encodings_are_independent(self):
        """Test each encoding carries its own values, not a shared buffer's."""
        first = Generic3AxisPedalOutput(left_pedal=0.5, right_pedal=0.3, rudder=0.2)
        second = Generic3AxisPedalOutput(left_pedal=0.9, right_pedal=0.7, rudder=-0.8)

        assert first.left_pedal == pytest.approx(0.5)
        assert first.right_pedal == pytest.approx(0.3)
        assert first.rudder == pytest.approx(0.2)
        assert second.left_pedal == pytest.approx(0.9)
        assert second.right_pedal == pytest.approx(0.7)
        assert second.rudder == pytest.approx(-0.8)


class TestGeneric3AxisPedalOutputEncoding:
    """Tests that an encoded payload reads back.

    A tracker with no pedal data returns None rather than an empty payload, so
    absence needs no case here; the source-node tests cover feeding None through.
    """

    def test_encoded_payload_reads_back(self):
        """An encoded payload gates as True and its fields read directly."""
        output = Generic3AxisPedalOutput(left_pedal=0.8, right_pedal=0.2, rudder=-0.5)

        assert output
        assert output.left_pedal == pytest.approx(0.8)
        assert output.right_pedal == pytest.approx(0.2)
        assert output.rudder == pytest.approx(-0.5)

    def test_repr_present(self):
        """Repr of a present payload names the type."""
        assert "Generic3AxisPedalOutput" in repr(Generic3AxisPedalOutput())


class TestGeneric3AxisPedalOutputRecordTimestamp:
    """Tests for Generic3AxisPedalOutputRecord with DeviceDataTimestamp."""

    def test_construction_with_timestamp(self):
        """Test Generic3AxisPedalOutputRecord carries DeviceDataTimestamp."""
        data = Generic3AxisPedalOutput(left_pedal=0.8, right_pedal=0.2, rudder=0.5)
        ts = DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        record = Generic3AxisPedalOutputRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data.left_pedal == pytest.approx(0.8)
        assert record.data.right_pedal == pytest.approx(0.2)
        assert record.data.rudder == pytest.approx(0.5)

    def test_payload_less_record(self):
        """A record may carry a timestamp and no payload: MCAP's frame sentinel."""
        record = Generic3AxisPedalOutputRecord(None, DeviceDataTimestamp(1, 2, 3))
        assert record.data is None
        assert record.timestamp.available_time_local_common_clock == 1

    def test_timestamp_fields(self):
        """Test all three DeviceDataTimestamp fields are accessible."""
        data = Generic3AxisPedalOutput()
        ts = DeviceDataTimestamp(111, 222, 333)
        record = Generic3AxisPedalOutputRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 111
        assert record.timestamp.sample_time_local_common_clock == 222
        assert record.timestamp.sample_time_raw_device_clock == 333
